File: sofascore-bridge.js
#!/usr/bin/env node
/**
 * SofaScore Python Wrapper Bridge
 * ===============================
 * 
 * Calls the Python sofascore-wrapper for maximum reliability
 * Then returns data in Node.js format
 * 
 * Usage:
 *   const { fetchWithPython } = require('./sofascore-bridge.js');
 *   const matches = await fetchWithPython();
 */

import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * Fetch matches using Python sofascore-wrapper
 * 
 * @returns {Promise<Array>} Array of match objects
 */
export async function fetchWithPython() {
  return new Promise((resolve, reject) => {
    const pythonScript = join(__dirname, 'sofascore-bridge.py');
    
    const python = spawn('python3', [pythonScript], {
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    let stdout = '';
    let stderr = '';

    python.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    python.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          resolve(result.matches || []);
        } catch (e) {
          console.error('Failed to parse Python output:', e.message);
          reject(e);
        }
      } else {
        if (stderr.includes('ModuleNotFoundError') || stderr.includes('sofascore-wrapper')) {
          reject(new Error(
            'sofascore-wrapper not installed. Install with: pip install sofascore-wrapper'
          ));
        } else {
          reject(new Error(`Python script error: ${stderr}`));
        }
      }
    });

    python.on('error', (err) => {
      reject(new Error(`Failed to spawn Python process: ${err.message}`));
    });
  });
}

/**
 * Try to fetch with Python, fallback to direct API
 */
export async function fetchWithFallback() {
  try {
    console.log('🐍 Trying Python wrapper (most reliable)...');
    const matches = await fetchWithPython();
    
    if (matches && matches.length > 0) {
      console.log(`✅ Python wrapper returned ${matches.length} matches`);
      return matches;
    }
  } catch (error) {
    console.warn('⚠️ Python wrapper error:', error.message);
    console.log('Falling back to direct API...');
  }

  // Fallback to direct API
  try {
    console.log('📡 Trying direct SofaScore API...');
    const matches = await fetchDirectAPI();
    
    if (matches && matches.length > 0) {
      console.log(`✅ Direct API returned ${matches.length} matches`);
      return matches;
    }
  } catch (error) {
    console.error('❌ Direct API error:', error.message);
  }

  return [];
}

/**
 * Fetch directly from SofaScore API
 */
async function fetchDirectAPI() {
  const sports = ['football', 'basketball', 'tennis', 'volleyball'];
  const allMatches = [];

  for (const sport of sports) {
    try {
      const response = await fetch(
        `https://api.sofascore.com/api/v1/sport/${sport}/events/last`,
        {
          headers: {
            'Accept': 'application/json',
            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Referer': 'https://www.sofascore.com/',
          },
        }
      );

      if (response.ok) {
        const data = await response.json();
        const events = data.events || [];

        const matches = events.map(event => ({
          match_id: `sofascore_${sport}_${event.id}`,
          sport: sport,
          league: `${event.tournament.category.name} - ${event.tournament.name}`,
          home_team_name: event.homeTeam.name,
          away_team_name: event.awayTeam.name,
          status: event.status.type === 'inprogress' ? 'live' 
                : event.status.type === 'finished' ? 'finished'
                : 'upcoming',
          home_score: (event.status.type === 'finished' || event.status.type === 'inprogress')
            ? event.homeScore.display 
            : null,
          away_score: (event.status.type === 'finished' || event.status.type === 'inprogress')
            ? event.awayScore.display 
            : null,
          start_time: new Date(event.startTimestamp * 1000).toISOString(),
          minute: event.status.type === 'inprogress' 
            ? Math.floor(event.time.currentDisplaySeconds / 60)
            : null,
        }));

        allMatches.push(...matches);
      }
    } catch (error) {
      console.warn(`⚠️ ${sport} error:`, error.message);
    }
  }

  return allMatches;
}

/**
 * CLI Usage
 */
async function main() {
  console.log('🌟 SofaScore Bridge\n');

  const matches = await fetchWithFallback();

  if (matches.length > 0) {
    console.log(`\n📊 Found ${matches.length} matches:\n`);
    matches.slice(0, 5).forEach(match => {
      console.log(`  ${match.home_team_name} vs ${match.away_team_name}`);
      console.log(`  Sport: ${match.sport} | League: ${match.league}`);
      console.log(`  Status: ${match.status}\n`);
    });

    if (matches.length > 5) {
      console.log(`... and ${matches.length - 5} more matches`);
    }
  } else {
    console.log('⚠️ No matches found');
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
}

File: sofascore-bridge.py
#!/usr/bin/env python3
"""
SofaScore Python Wrapper Bridge
================================

Fetches real sports data using sofascore-wrapper and outputs JSON
for Node.js/JavaScript integration.

Installation:
    pip install sofascore-wrapper

Usage:
    python3 sofascore-bridge.py

Output:
    JSON array of matches
"""

#!/usr/bin/env python
import json
import sys
from datetime import datetime
from typing import List, Dict, Any

# Try to import sofascore
try:
    from sofascore import SofaScoreAPI
except ImportError as e:
    # Try alternative import paths
    try:
        import sofascore_wrapper
        from sofascore_wrapper import SofaScoreAPI
    except ImportError:
        print(
            json.dumps({
                "error": f"sofascore-wrapper not found. Install with: pip install sofascore-wrapper. Details: {str(e)}",
                "install": "pip install sofascore-wrapper",
                "matches": []
            })
        )
        sys.exit(1)


def fetch_sport_matches(api: SofaScoreAPI, sport: str) -> List[Dict[str, Any]]:
    """
    Fetch matches for a specific sport
    
    Args:
        api: SofaScoreAPI instance
        sport: Sport name (football, basketball, tennis, etc.)
    
    Returns:
        List of match dictionaries
    """
    try:
        events = api.get_sport_events(sport)
        matches = []
        
        for event in events:
            try:
                match = {
                    "match_id": f"sofascore_{sport}_{event.id}",
                    "sport": sport,
                    "league": f"{event.tournament.category.name} - {event.tournament.name}",
                    "home_team_name": event.home_team.name,
                    "away_team_name": event.away_team.name,
                    "status": (
                        "live" if event.status == "inprogress"
                        else "finished" if event.status == "finished"
                        else "upcoming"
                    ),
                    "home_score": event.home_score if event.home_score is not None else None,
                    "away_score": event.away_score if event.away_score is not None else None,
                    "start_time": event.start_timestamp.isoformat() if hasattr(event.start_timestamp, 'isoformat') else str(event.start_timestamp),
                    "minute": int(event.minute) if event.minute is not None else None,
                }
                matches.append(match)
            except Exception as e:
                print(f"Warning: Error processing event {event.id}: {e}", file=sys.stderr)
                continue
        
        return matches
    
    except Exception as e:
        print(f"Error fetching {sport}: {e}", file=sys.stderr)
        return []


def main():
    """Main entry point"""
    try:
        api = SofaScoreAPI()
        
        # Fetch multiple sports
        sports = ["football", "basketball", "tennis", "volleyball"]
        all_matches = []
        
        print(f"Fetching from SofaScore...", file=sys.stderr)
        
        for sport in sports:
            matches = fetch_sport_matches(api, sport)
            all_matches.extend(matches)
            print(f"✓ {sport}: {len(matches)} matches", file=sys.stderr)
        
        # Output JSON
        result = {
            "success": True,
            "count": len(all_matches),
            "timestamp": datetime.now().isoformat(),
            "matches": all_matches
        }
        
        print(json.dumps(result, indent=2))
    
    except Exception as e:
        error_result = {
            "success": False,
            "error": str(e),
            "matches": []
        }
        print(json.dumps(error_result, indent=2), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
